internals-js/src/operations.ts

```diff
@@ -934,25 +936,41 @@ export class Operation extends DirectiveTargetElement<Operation> {
       this.appliedDirectives,
     );
   }
+  
+  private collectVariablesFromFragments(allAvailableVariables: VariableDefinitions, fragments: NamedFragments): VariableDefinitions {
+    const varDefs = new VariableDefinitions();
+    varDefs.addAll(this.variableDefinitions);
+
+    const collector = new VariableCollector();
+    collectVariablesInNamedFragments(fragments, collector);
+    
+    for (const v of collector.variables()) {
+      const def = allAvailableVariables.definition(v.name);
+      if (def) {
+        varDefs.add(def);
+      }
+    }
+    return varDefs;
+  }
 
   // Returns a copy of this operation with the provided updated selection set and fragments.
-  private withUpdatedSelectionSetAndFragments(newSelectionSet: SelectionSet, newFragments: NamedFragments | undefined): Operation {
+  private withUpdatedSelectionSetAndFragments(newSelectionSet: SelectionSet, newFragments: NamedFragments | undefined, allAvailableVariables?: VariableDefinitions): Operation {
     if (this.selectionSet === newSelectionSet && newFragments === this.fragments) {
       return this;
     }
-
+    
     return new Operation(
       this.schema(),
       this.rootKind,
       newSelectionSet,
-      this.variableDefinitions,
+      (allAvailableVariables && newFragments) ? this.collectVariablesFromFragments(allAvailableVariables, newFragments) : this.variableDefinitions,
```

Comment:
Note that this makes this method subtly incorrect, as it will omit the selection set's fragments in the variable definitions.